feat(report-portal): plugin implementation

yashoswalyo opened the pull request:
### Changes
- added Instances Page
- added Projects Page
- added Launches Page
- Updated Report Portal API
- Added api pagination for table
- fixed tsc and lint issues in backend plugin
- added Readme files for both plugins

### Screenshots
- Report Portal Global page / Instances Page
  ![image](https://github.com/appdev-platform/janus-backstage-plugins/assets/42431299/ba131364-d5b8-423e-8d5a-521c330d0a57)
- Projects Page
  ![image](https://github.com/appdev-platform/janus-backstage-plugins/assets/42431299/cc48f44b-ac30-4fd4-a4ae-f006d6d87e8f)
- Launches Page
  ![image](https://github.com/appdev-platform/janus-backstage-plugins/assets/42431299/576e900d-67eb-421c-8779-11bfb5b3091a)
- Test Statistics Card
  ![image](https://github.com/appdev-platform/janus-backstage-plugins/assets/42431299/bae18643-6976-425c-8bc8-db2448d712d6)

riginoommen commented on plugins/report-portal/README.md:
````diff
@@ -11,3 +11,83 @@ Your plugin has been added to the example app in this repository, meaning you'll
+- Run the following command in your backstage project
+
+  ```shell
+  yarn workspace app add @janus-idp/report-portal
````
```suggestion
  yarn workspace app add @appdev-platform/backstage-plugin-report-portal
```

plugin name should be `@appdev-platform/backstage-plugin-report-portal`

riginoommen replied:
update in all places

riginoommen commented on plugins/report-portal/README.md:
````diff
@@ -11,3 +11,83 @@ Your plugin has been added to the example app in this repository, meaning you'll
+  - open `/packages/app/src/App.tsx` and add the following code
+
+    ```js
+    import { ReportPortalGlobalPage } from '@janus-idp/backstage-plugin-report-portal';
````
same as above

riginoommen commented on plugins/report-portal/package.json:
```diff
@@ -0,0 +1,76 @@
+  "configSchema": "config.d.ts",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/janus-idp/backstage-plugins",
```
use appdev platform repo urls

yashoswalyo replied:
this will be released on npm under janus-idp, still should we change repo-url?

riginoommen replied:
We have to move to janus idp till a community repo open up till that time we will be maintaining it

riginoommen commented on plugins/report-portal/src/api/ReportPortalClient.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+    baseUrl.searchParams.append('host', host);
+    const response = await fetch(baseUrl);
+    if (response.status >= 400 && response.status < 600) {
```
make a condion of response.status != 200

deshmukhmayur replied:
It should be an inverted condition for `>= 200` && `< 300`

riginoommen commented on plugins/report-portal/src/api/ReportPortalClient.ts:
```diff
@@ -0,0 +1,66 @@
+    const baseUrl = new URL(`project/${projectId}`, await this.getBaseApiUrl());
+    baseUrl.searchParams.append('host', host);
+    const response = await fetch(baseUrl);
+    if (response.status >= 400 && response.status < 600) {
```
make a condion of response.status != 200

yashoswalyo replied:
changed this condition

riginoommen commented on plugins/report-portal/src/api/ReportPortalClient.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+    baseUrl.searchParams.append('host', host);
+    const response = await fetch(baseUrl);
+    if (response.status >= 400 && response.status < 600) {
```
make a condion of response.status != 200

riginoommen commented on plugins/report-portal/src/api/types.ts:
```diff
@@ -1,6 +1,15 @@
+  usersQuantity: number;
+  id: number;
+  launchesQuantity: number;
+  launchesPerUser: null;
```
define proper data types for the types

yashoswalyo replied:
updated the type definitions

riginoommen commented on plugins/report-portal/src/components/LaunchesPage/LaunchesPage.tsx:
```diff
@@ -0,0 +1,63 @@
+import { LaunchesPageContent } from './LaunchesPageContent/LaunchesPageContent';
+
+export const LaunchesPage = () => {
+  const rootR = useRouteRef(rootRouteRef);
```
```suggestion
  const rootR = useRouteRef(rootRouteRef);
```

use complete readable variables

yashoswalyo replied:
changed the var. names here

riginoommen commented on plugins/report-portal/src/components/LaunchesPage/LaunchesPage.tsx:
```diff
@@ -0,0 +1,63 @@
+        title={
+          <>
+            <Breadcrumbs>
+              <Link color="inherit" to={rootR()}>
```
```suggestion
              <Link color="inherit" to={rootR()}>
```

use complete readable variables

yashoswalyo replied:
changed the var. names here

riginoommen commented on plugins/report-portal/src/components/LaunchesPage/LaunchesPageContent/LaunchesPageContent.tsx:
```diff
@@ -0,0 +1,175 @@
+      .getLaunchResults(project, host, {
+        'page.size': 10,
+        'page.page': 1,
+        'page.sort': 'startTime,DESC',
```
```suggestion
        'page.sort': 'startTime, DESC',
```

yashoswalyo replied:
getting 400 response with that
````
{
  "errorCode":40012,
  "message":"Sorting parameter  DESC is not defined"
}

yashoswalyo replied:
@riginoommen can you resolve this?

riginoommen commented on plugins/report-portal/src/components/LaunchesPage/LaunchesPageContent/LaunchesPageContent.tsx:
```diff
@@ -0,0 +1,175 @@
+      .getLaunchResults(project, host, {
+        'page.size': pageSize,
+        'page.page': page + 1,
+        'page.sort': 'startTime,DESC',
```
```suggestion
        'page.sort': 'startTime, DESC',
```

yashoswalyo replied:
getting 400 response with that
````
{
  "errorCode":40012,
  "message":"Sorting parameter  DESC is not defined"
}

yashoswalyo replied:
this one too

riginoommen commented on plugins/report-portal/src/components/LaunchesPage/LaunchesPageContent/LaunchesPageContent.tsx:
```diff
@@ -0,0 +1,175 @@
+      title: 'Start Time',
+      align: 'center',
+      width: '25%',
+      render: row => moment(row.startTime).fromNow(),
```
moment support has be depricated use alternatives such as dayjs

riginoommen replied:
https://momentjs.com/docs/#/-project-status/

riginoommen commented on plugins/report-portal/src/components/ProjectsPage/ProjectsPage.tsx:
```diff
@@ -0,0 +1,54 @@
+import { ProjectsPageContent } from './ProjectsPageContent/ProjectsPageContent';
+
+export const ProjectsPage = () => {
+  const rootR = useRouteRef(rootRouteRef);
```
```suggestion
  const rootR = useRouteRef(rootRouteRef);
```

use complete/proper variables

riginoommen commented on plugins/report-portal/src/components/LaunchesPage/LaunchesPageContent/LaunchesPageContent.tsx:
```diff
@@ -0,0 +1,175 @@
+
+import { IconButton, Link } from '@material-ui/core';
+import Launch from '@material-ui/icons/Launch';
+import moment from 'moment';
```
https://momentjs.com/docs/#/-project-status/

try considering other libraries like dayjs

deshmukhmayur replied:
Moment is deprecated and no longer maintained, and they recommend using other libraries. As @riginoommen mentioned, `dayjs` is a good alternative, or consider using `luxon` for advanced features.

riginoommen commented on plugins/report-portal-backend/README.md:
````diff
@@ -12,3 +12,64 @@ start` in the root directory, and then navigating to [/report-portal](http://loc
+
+- Install the plugin
+  ```shell
+  yarn workspace backend add @janus-idp/backstage-plugin-report-portal-backend
````
use appdevplatform namespace

deshmukhmayur commented on plugins/report-portal/src/plugin.ts:
```diff
@@ -0,0 +1,47 @@
+
+export const ReportPortalGlobalPage = reportPortalPlugin.provide(
+  createRoutableExtension({
+    name: 'RportPortalGlobalPage',
```
```suggestion
    name: 'ReportPortalGlobalPage',
```

yashoswalyo replied:
fixed the typo

deshmukhmayur commented on plugins/report-portal/src/api/types.ts:
```diff
@@ -50,19 +58,25 @@ export type LaunchDetails = {
+  totalPages: number;
+};
+
 export type LaunchDetailsResp = {
```
Use the full names for types. `Resp` can be confusing for other. Same for the `LaunchDetailsResp`

```suggestion
export type LaunchDetailsResponse = {
```

yashoswalyo replied:
made the changes

deshmukhmayur commented on plugins/report-portal/src/utils/isReportPortalAvailable.ts:
```diff
@@ -0,0 +1,9 @@
+  return Boolean(
+    entity.metadata.annotations?.['reportportal.io/project-name'] &&
+      entity.metadata.annotations?.['reportportal.io/launch-name'] &&
+      entity.metadata.annotations?.['reportportal.io/host'],
```
The host annotation should be optional. And if it's not provided, the plugin should use the first available integration from the app-config.

Since in most cases, there will only be 1 report portal integration. And in case they have multiple, they can add the host annotations as needed.

yashoswalyo replied:
removed it from here, and added check in ReportPortalCard component

deshmukhmayur commented:
Also, @yashoswalyo , the build is failing due to some changes in the yarn.lock. Can you update the yarn.lock in your branch?

yashoswalyo commented:
Ok

yashoswalyo committed:
feat(report-portal): add global page for report portal plugin

- added Instances Page
- added Projects Page
- added Launches Page
- Updated Report Portal API
- Added api pagination for table
- fixed tsc and lint issues in backend plugin
- added Readme files for both plugins

Signed-off-by: Yash Oswal <yashoswal18@gmail.com>
